ramlfications/types.py

```diff
@@ -0,0 +1,192 @@
+
+    def validate(self, validated_objet, position_hint=None):
+        """Basic default validator which does not check anything
+        :param <any> validated_objet: object to be validated with the
+            type definition
+        :param string position_hint: position of the object in a
+            more global validation. Used for proper error message
+        """
+        pass
+
+
+###
+#   helpers for ObjectType
+###
+@attr.s
+class Property(object):
+    """
+    helper class for holding additional checks for object properties
+
+    :param bool required: is this property mandatory
+    :param <any> default: default value for this property
+    :param <BaseType> data_type: type definition of the property
+    """
+    required        = attr.ib(default=None, repr=False)
+    default         = attr.ib(default=None, repr=False)
+    data_type       = attr.ib(default=None)
+
+
+def create_property(property_def):
+    # we remove the attributes for the property in order to keep
+    # only attributes needed for the type definition
+    type_def = copy.copy(property_def)
+    return Property(default=type_def.pop('default', None),
+                    required=type_def.pop('required', None),
+                    data_type=create_type(None, type_def))
+
+
+def parse_properties(properties):
+    # @todo: should parse k for syntax sugar
+    return OrderedDict([
+        (k, create_property(v))
+        for k, v in iteritems(properties)])
+
+
+@type_class("object")
+@attr.s
+class ObjectType(BaseType):
+    """
+    Type class for object types
+
+    :param string properties: dictionary of Properties
+    """
+    properties           = attr.ib(default=None, convert=parse_properties)
+    # to be implemented
+    min_properties        = attr.ib(repr=False, default=0)
+    max_properties        = attr.ib(repr=False, default=MAXSIZE)
+    additional_properties = attr.ib(repr=False, default=None)
+    pattern_properties    = attr.ib(repr=False, default=None)
+    discriminator        = attr.ib(repr=False, default=None)
+    discriminator_value   = attr.ib(repr=False, default=None)
+
+    def validate(self, o, position_hint=None):
+        if not isinstance(o, dict):
+            raise DataTypeValidationError(
+                position_hint, o,
+                "requires a dictionary")
+        if position_hint is None:
+            position_hint = ['object']
+
+        for k, p in iteritems(self.properties):
+            if p.required and k not in o:
+                raise DataTypeValidationError(
+                    position_hint + [k], None,
+                    "should be specified")
+            v = o.get(k, p.default)
+            p.data_type.validate(v, position_hint + [k])
+
+
+###
+#   helpers for StringType
+###
+def maybe_create_re(pattern):
+    if pattern is not None:
+        return re.compile(pattern)
+    return None
+
+
+@type_class("string")
+@attr.s
+class StringType(BaseType):
```

Comment:
Looking at the [spec diagram](http://docs.raml.org/specs/1.0/#raml-10-spec-overview), I feel like there should be four "base" types, `ObjectType`, `ArrayType`, `ScalarType`, and `ExternalType`, all of which inherit from `BaseType`. Then `ScalarType` would have `facets` and `enum`.  Then there would be `StringType` or `StringScalarType` that would inherit from `ScalarType`.